rc/tools/lint.kak

```diff
@@ -1,113 +1,339 @@
+
+        Assumes selections all have anchor before cursor, and that
+        %val{selections} and %val{selections_desc} are in the same order.
+    } \
+    lint-cleaned-selections \
+%{
+    # Clear the current contents of the various options.
+    set-option buffer lint_flags %val{timestamp}
+    set-option buffer lint_messages %val{timestamp}
+    set-option buffer lint_error_count 0
+    set-option buffer lint_warning_count 0
+
+    # Create a temporary directory to keep all our state.
     evaluate-commands %sh{
-        if [ -z "${kak_opt_lintcmd}" ]; then
-            echo 'fail The `lintcmd` option is not set'
-            exit 1
-        fi
+        # This is going to come in handy later.
+        kakquote() {
+            printf "'"
+            printf "%s" "$*" | sed "s/'/''/g"
+            printf "'"
+        }
+
+        # Before we clobber our arguments,
+        # let's record the lintcmd we were given.
+        lintcmd="$1"
 
+        # Some linters care about the name or extension
+        # of the file being linted, so we'll store the text we want to lint
+        # in a file with the same name as the original buffer.
         filename="${kak_buffile##*/}"
 
+        # A directory to keep all our temporary data.
         dir=$(mktemp -d "${TMPDIR:-/tmp}"/kak-lint.XXXXXXXX)
-        mkfifo "$dir"/fifo
-        printf '%s\n' "evaluate-commands -no-hooks write -sync $dir/${filename}"
 
+        # A fifo to send the results back to a Kakoune buffer.
+        # FIXME: Should we put the lint output in toolsclient?
+        mkfifo "$dir"/fifo
         printf '%s\n' "evaluate-commands -draft %{
-                  edit! -fifo $dir/fifo -debug *lint-output*
+                  edit! -fifo $(kakquote "$dir/fifo") -debug *lint-output*
                   set-option buffer filetype make
                   set-option buffer make_current_error_line 0
-                  hook -always -once buffer BufCloseFifo .* %{ nop %sh{ rm -r '$dir' } }
               }"
 
+        # Write all the selection descriptions to files.
+        eval set -- "$kak_quoted_selections_desc"
+        i=0
+        for desc; do
+            mkdir -p "$dir"/sel-"$i"
+            printf "%s" "$desc" > "$dir"/sel-$i/desc
+            i=$(( i + 1 ))
+        done
+
+        # Write all the selection contents to files.
+        eval set -- "$kak_quoted_selections"
+        i=0
+        for text; do
+            # The selection text needs to be stored in a subdirectory,
+            # so we can be sure the filename won't clash with one of ours.
+            mkdir -p "$dir"/sel-"$i"/text/
+            printf "%s" "$text" > "$dir"/sel-$i/text/"$filename"
+            i=$(( i + 1 ))
+        done
+
+        # We do redirection trickiness to record stderr from
+        # this background task and route it back to Kakoune,
+        # but shellcheck isn't a fan.
+        # shellcheck disable=SC2094
         ({ # do the parsing in the background and when ready send to the session
```

Comment:
The extra layer was added in 009eb80. Apparently OpenBSD's `/bin/sh` is buggy and cannot run a command in the background if its output is redirected at the same time.

Reply:
Could it be that the commands that use that (not OpenBSD-proof) pattern are broken on OpenBSD, then? It caught my eye because it's the only place I am seeing it.

Reply:
The [original PR](https://github.com/mawww/kakoune/pull/2955) says:

> I'm not sure if the `(...) > /dev/null 2>&1 < /dev/null &` pattern should be adapted in all other places as well. The tools using fifos were just the first time I experienced problems with it.

It's not exactly clear to me how fifos need to be involved in order to tickle the bug, but I assume the other uses don't involve fifos, or they use fifos in a safe way, or they just weren't plugins that @codesoap happened to use so he didn't realise they needed to be fixed.

Reply:
Just in case you didn't already find it: I condensed the problem in [a mail](https://marc.info/?l=openbsd-misc&m=156033323516331&w=2) to the OpenBSD community.